Add GKE support to gcp-service-discovery

stephen-soltesz opened the pull request:
This change adds support for collecting targets from annotated services defined in GKE clusters. This allow us to automatically discover and collect metrics from prometheus instances running in other gke clusters via "prometheus federation"

To support the generation of various `TargetSource` types, this change also adds a `Factory` interface that returns new `TargetSource` instances.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/gcp-service-discovery/3)
<!-- Reviewable:end -->

gfr10598 commented on aeflex/aeflex.go:
```diff
@@ -49,9 +53,19 @@ type AEFlexSource struct {
 
 // NewAEFlexSource returns a new discovery object with an authenticated AppEngine client.
-func NewAEFlexSource(project string) (*AEFlexSource, error) {
+func NewAEFlexSource(project, filename string) *AEFlexSource {
```
Maybe call this NewDiscoverySource?  AEFlex is redundant.  I see that means changing the struct name as well.  8-(

gfr10598 commented on aeflex/aeflex.go:
```diff
@@ -49,9 +53,19 @@ type AEFlexSource struct {
+}
+
+// Client creates a new initialized AEFlexSource.
+func (client *AEFlexSource) Client() (targetsource.TargetSource, error) {
```
"client" seems confusing here.

gfr10598 commented on aeflex/aeflex.go:
```diff
@@ -49,9 +53,19 @@ type AEFlexSource struct {
+
+// Client creates a new initialized AEFlexSource.
+func (client *AEFlexSource) Client() (targetsource.TargetSource, error) {
+	d := &AEFlexSource{
```
Add comment explaining why you need a new AEFlexSource instead of just using client?

Also, could this just be copied using copy or assignment? d := *client ?

gfr10598 commented on aeflex/aeflex.go:
```diff
@@ -132,7 +146,7 @@ func (client *AEFlexSource) getLabels(service *appengine.Service, version *appen
 
 // Save writes the content of the the collected set of targets.
-func (client *AEFlexSource) Save(filename string) error {
+func (client *AEFlexSource) Save() error {
```
s/client/source/ ?
s/client/src/ ?

stephen-soltesz commented:
PTAL?

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[aeflex/aeflex.go, line 56 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-56:-Kjs2Br2exKUai22n8Pe:b3lntms) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L56)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Maybe call this NewDiscoverySource?  AEFlex is redundant.  I see that means changing the struct name as well.  8-(
</blockquote></details>

What would be the name of the corresponding call in the gke package? I think it's common to name `New` functions after the type. Maybe the `TargetSource` interface should be renamed to `DiscoverySource`?

---

*[aeflex/aeflex.go, line 65 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-65:-Kjs1jhFlBx8n6_bdNmd:b-r8lgsq) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L65)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

"client" seems confusing here.
</blockquote></details>

Renamed to `aef`.

---

*[aeflex/aeflex.go, line 66 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-66:-Kjs2vP_A4evLFD_RMWn:bl4h8b4) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L66)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Add comment explaining why you need a new AEFlexSource instead of just using client?

Also, could this just be copied using copy or assignment? d := *client ?
</blockquote></details>

Uuuh. That's a really good question. I'm not sure we need the second object now. Since there were two interfaces, I was probably thinking that I needed two objects, and I wanted to avoid defining a tiny intermediate type just for that. But, I think we just have a case where a single struct can implement two interfaces nicely.

How does it look?

---

*[aeflex/aeflex.go, line 149 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-149:-Kjs1puL70xYxUM89Jx0:b-r8lgsq) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L149)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

s/client/source/ ?
s/client/src/ ?
</blockquote></details>

Renamed to `aef`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-:-Kjs4CL0iF67y3eOQAUG:b-833305)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
PTAL?

Based on in-person discussion, I've made significant renames throughout to help clarify the intention of the new and existing interfaces.

How does it look?

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-:-KjsT9ezH2Buw1TzgWa2:bgect98)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[aeflex/aeflex.go, line 56 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-56:-KjtL6kNQJpbWWNadVF-:byou5pc) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L56)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What would be the name of the corresponding call in the gke package? I think it's common to name `New` functions after the type. Maybe the `TargetSource` interface should be renamed to `DiscoverySource`?
</blockquote></details>

Discussed offline and reflected in code now.

---

*[aeflex/aeflex.go, line 65 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-65:-KjtL1fQvwfVG4Tp8og9:ba5xp6o) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L65)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Renamed to `aef`.
</blockquote></details>

Comment obsolete?

---

*[aeflex/aeflex.go, line 66 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-66:-KjtKmquhVWRGs3xHx4t:bao9msw) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L66)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Uuuh. That's a really good question. I'm not sure we need the second object now. Since there were two interfaces, I was probably thinking that I needed two objects, and I wanted to avoid defining a tiny intermediate type just for that. But, I think we just have a case where a single struct can implement two interfaces nicely.

How does it look?
</blockquote></details>

Discussed offline separating Factory from Source...

---

*[aeflex/aeflex.go, line 149 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-149:-KjtLGkDcY1WLzoMeB-9:ba5xp67) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L149)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Renamed to `aef`.
</blockquote></details>

Comment obsolete.

---

*[aeflex/aeflex.go, line 72 at r2](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjtHCxljpTX5lWHLR9_:-KjtHCxljpTX5lWHLR9a:bgovn8r) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/642f0b1c50bd42c3640218c4f88fb58f390a96e6/aeflex/aeflex.go#L72)):*
> ```Go
> // Create returns a discovery.Source initialized with authenticated clients for
> // App Engine Admin API, ready for Collection.
> func (f *Factory) Create() (discovery.Source, error) {
> ```

FYI, I've been usually keeping the functions near the struct that they belong to.  I don't know that this is a useful convention, but would appreciate your thoughts on it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-:-KjtLDE1NdnVVgchv7wy:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
I've moved the Source struct closer to the functions that use them.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[aeflex/aeflex.go, line 65 at r1](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjoYGmcT4YThNpxr2-O-r1-65:-Kjta6Lgyx3hM0ePL0qs:b6igr6n) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/9df8773082f774b41fd5eac85b3ce5793115ac76/aeflex/aeflex.go#L65)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Comment obsolete?
</blockquote></details>

Yes. All fixed based on our in-person conversation.

---

*[aeflex/aeflex.go, line 72 at r2](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-KjtHCxljpTX5lWHLR9_:-KjtbqzAFiBZoJZWMnl4:b-ejxawo) ([raw file](https://github.com/m-lab/gcp-service-discovery/blob/642f0b1c50bd42c3640218c4f88fb58f390a96e6/aeflex/aeflex.go#L72)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

FYI, I've been usually keeping the functions near the struct that they belong to.  I don't know that this is a useful convention, but would appreciate your thoughts on it.
</blockquote></details>

Good question. I think I picked up the habit of putting all the structs together from maintaining proto service definitions in my former life. Godoc doesn't seem to care and automatically groups struct and related functions together regardless of the code. So, doing that in the code also makes sense. Let's try that.

I've moved the Source struct closer to the functions below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/gcp-service-discovery/3#-:-KjteqVbmFlsiUvexQSo:b-l5xhp9)*
<!-- Sent from Reviewable.io -->
